### ISSUE TITLE
Map key checker does not properly perform type refinement for local variables

### ISSUE THREAD

**Issue by atomicknight** [typetools/checker-framework#514]
Environment: Checker Framework 1.9.8, Oracle JDK 1.8.0_65, Windows 7

Given this class:
```java
public class EntrySet {
	public static void main(String[] args) {
		Map<String, String> m = new HashMap<>();
		Set<Map.Entry<String, String>> es = m.entrySet();
	}
}
```

The following error is raised:
```
d:\workspace\cf-testcases\mapkey2\src\main\java\org\example\EntrySet.java:10: error: [assignment.type.incompatible] incompatible types in assignment.
		Set<Map.Entry<String, String>> es = m.entrySet();
		                                              ^
  found   : @UnknownKeyFor Set<@UnknownKeyFor Entry<@KeyFor("m") String, @UnknownKeyFor String>>
  required: @UnknownKeyFor Set<@UnknownKeyFor Entry<@UnknownKeyFor String, @UnknownKeyFor String>>
```

The full test case (Maven build - just execute `mvn`): https://github.com/atomicknight/cf-testcases/tree/master/mapkey2

**Comment by smillst** [comment]
This is expected behavior, because
@UnknownKeyFor Set<@UnknownKeyFor Entry<@KeyFor("m") String, @UnknownKeyFor String>>
is not a subtype of 
@UnknownKeyFor Set<@UnknownKeyFor Entry<@UnknownKeyFor String, @UnknownKeyFor String>>
  See http://types.cs.washington.edu/checker-framework/current/checker-framework-manual.html#type-refinement-types-that-are-not-refined.   

I'm going to close this issue, but let us know if you have questions about this.

**Comment by atomicknight** [comment]
Thanks for the pointer to the documentation. So if I understand correctly, every local variable must be explicitly annotated? If so, would it make sense to add a note to the documentation under the "Map Key Checker" section?

**Comment by smillst** [comment]
Local variables in general do not require annotations. 
```
@KeyFor("m") String k;
String copyK = k; 
//The type of copyK is refined from @UnknownKey String to @KeyFor("m")
```

 Local variables that have a generic type may require an annotation on that generic type in order for an assignment to type check.

```
List<@KeyFor("m") String> a = ...;
List<String> b = a; // error: List<@KeyFor(m) String> is not a subtype of List<@UnknownKeyFor String>
```
Does that make sense?

**Comment by atomicknight** [comment]
Yes, that makes sense. I just wanted to confirm that invoking either of `Map#keySet` or `Map#entrySet` will always require the local variable to be explicitly annotated because the return types of those methods include type parameters annotated with `@KeyFor`. Seems like that is the case?

**Comment by smillst** [comment]
Yes, that's correct.

**Comment by wmdietl** [comment]
I thought we already have an issue for this, but I can't find it: at the moment type refinement is only for the top-level qualifier. If the local variable is effectively final, we should be able to also refine generic types. In situations like the "keySet()" calls, this would be very helpful. If there is no issue for this, should we create one?

**Comment by mernst** [comment]
Yes, we could create an enhancement request for this.

**Comment by smillst** [comment]
@wmdietl Is Issue #260 the one you are thinking of?
